### ISSUE TITLE
Suggestion: Crafting thresholds for fur product crafting

### ISSUE THREAD

**Issue by Quizer85** [PrincessRTFM/AutoKittens#11]
The crafting threshold implementation you came up with should also be used for fur products, and it would in fact be ideal for that. Every step in that chain is typically limited by input products, not the culture or science needed to upgrade it.

Each tier should have one of those dropdown menus that lets you set what percentage of the previous tier to maintain. Unlike the complex web of the other resources, this should not typically cause unforeseen / undesirable behavior.

**Comment by VariableVixen** [comment]
Hm. I think there may be a misunderstanding of how the system works. The loop that handles actually _doing_ the autocrafting has been rewritten to use this new system. _Any_ threshold-based autocrafting from this script follows it, and it dynamically determines what type of condition to apply based on the input values. The relevant logic is almost entirely on [lines 1477 to 1496](https://github.com/PrincessRTFM/AutoKittens/blob/5dccb1e6b811ee81113153f31e540ff8690ebc92/AutoKittens.js#L1477-L1496), but you might also want to look around it. Every output product has its input ingredients checked through that logic. In particular, line 1480 applies a check to resources that _have_ a cap, and line 1487 applies a different check - the new one - to ones that _don't_.

Bottom line, fur product crafting already uses the same updated (technically hybridised) crafting system that everything else being autocrafted does. The _before hunting_ options are separate, but they also specify that they craft _all_ (the output that they can) before the hunt.

**Comment by Quizer85** [comment]
Ugh... if it's like that, how are we going to get separate thresholds? I can see only having one percentage setting for the whole fur product chain, but the other resources really call for individual settings.

I guess you could have an array of threshold values, with each of them paired to a type of crafting. The crafting loop knows which one it is doing, right? So it would just have to fetch the correct percentage value according to whatever `i` it is on.

**Comment by VariableVixen** [comment]
It'll be easy to implement separate thresholds per-input _or_ per-output, but trying to do both... I'm not sure about that one. That would be... _very_ messy. So I can do "when crafting _X_ use threshold _Y_ for all uncapped inputs" or I could do "when considering uncapped input _X_ use threshold _Y_" but not really both.

**Comment by Quizer85** [comment]
Hm... what exactly IS the crafting loop looking at? It does something like "for each`i`, look at condition and if it is met, craft". But are the `i` basically crafting recipes (outputs), or are they the inputs?

How many recipes are there that have multiple storage-unlimited inputs? Ships, tankers and megaliths are the ones I can see. All of these use multiple storage-unlimited inputs and no storage-limited inputs. Maybe we should just special-case these. How hard would it be to have something like this for each of them?
![image](https://user-images.githubusercontent.com/14056473/62195171-b628fe80-b37b-11e9-8a9a-50d03a32f7ba.png)

**Comment by VariableVixen** [comment]
The loop is actually a nested loop - for each craftable (that is unlocked and enabled) it loops over each input resource. If it's a capped one, it checks one way; uncapped are checked the same as before. _All_ ingredients for a given output are automatically checked in this way. If _any_ of the checks fail (not enough of the ingredient according to whatever formula is being applied) it skips that craft entirely, moving immediately to the next one.

```
for each RECIPE in CRAFTING_RECIPES:
	if RECIPE is disabled, skip this RECIPE
	for each INGREDIENT in RECIPE INPUTS:
		if the CURRENT AMOUNT of this INGREDIENT is too low to actually craft this RECIPE, skip this RECIPE
		if INGREDIENT is capped:
			if the CURRENT AMOUNT of INGREDIENT is below the STORAGE THRESHOLD, skip this RECIPE
		if INGREDIENT is not capped:
			calculate the RATIO PERCENTAGE of the CURRENT AMOUNT of the RECIPE OUTPUT divided by the CURRENT AMOUNT of this INGREDIENT
			if the RATIO PERCENTAGE is above the SECONDARY CRAFTING THRESHOLD, skip this RECIPE
	if we get here, then all of the ingredient level checks are passed, so CRAFT this RECIPE
```

It'd be easy enough to set up separate thresholds for separate things, but only to a certain point - for instance, "when crafting concrete, use this threshold for <capped/uncapped> inputs" _or_ "when crafting _with_ steel, use this threshold _for steel as an input_" but I can't really mix the two, so no "when crafting concrete, use _X_ as the threshold for steel and use _Y_ as the threshold for slabs, but when crafting alloy use _Z_ as the threshold for steel instead" - that would be _far_ too messy to try and implement. I could _maybe_ find a way to allow you to mix the two in the sense of "the steel threshold is normally _X_ but whenever you craft alloy _all_ uncapped resources use the threshold _Y_ instead" but that'd be about the extent of it.

**Comment by Quizer85** [comment]
So, what I would naively do here is this:
```
if INGREDIENT is not capped:
	if INGREDIENT does not matter for this RECIPE, skip this INGREDIENT
	if INGREDIENT does matter for this RECIPE:
		calculate the RATIO PERCENTAGE of the CURRENT AMOUNT of the RECIPE OUTPUT divided by the CURRENT AMOUNT of this INGREDIENT
		if the RATIO PERCENTAGE is above the SECONDARY CRAFTING THRESHOLD, skip this RECIPE
```
Seems pretty straightforward. The user sets a bunch of variables via the options that determine which ingredients matter for which recipes. If an ingredient doesn't matter, the script stops looking at that ingredient and moves on to the next one for that recipe.

I'm pretty sure this would currently only be needed in the `INGREDIENT is not capped` part of the loop, but if you wanted to be super thorough and future-proof, I guess you could implement it for `INGREDIENT is capped` as well. (It would have the beneficial side effect that the user could selectively ignore the storage threshold and craft specific recipes as soon as enough resources are available.)

The problem I suspect you're about to tell me is that `for this RECIPE` is the difficult part. When the script is handling CONCRETE, it fetches the ingredient STEEL, which is not in any way distinguishable from when later on the script gets around to ALLOY, which also fetches STEEL. Is that it?

**Comment by VariableVixen** [comment]
You'd loop _every_ resource in the game, by doing that. This loop only examines the inputs for the current recipe, your implementation would have to look at every resource in the game, for every crafting recipe in the game.

This implementation does not (currently) differentiate "this is steel, and we're making concrete" from "this is steel, and we're making alloy", but it _could_. The problem is that to do that, every recipe would need another option for every input, _for every setting_ that needs to differentiate at that level. That'd currently would come out to 34 new options _per setting_, if I have that math right. More importantly, it would introduce a _lot_ of new bugs, because there are suddenly _more_ than 34 new points of failure - it has to set the options, it has to determine which options to pay attention to in a given loop, it has to read the options, and then it also has to apply them.

Within the _outer_ loop (the one that iterates over all of the crafts), these things are known or are derived/obtained from things that are known:
- the internal name (resource ID) of what is being crafted
- the resource object for what is being crafted
- the list (as an array) of internal names (resource IDs) of all of the things that this craft requires
- the quantity of each crafting input that is required for this craft

Within the _inner_ loop (that iterates over the inputs), the internal name (resource ID) _and_ resource object for the input being checked are also available. The resource objects are what provide the current/max values of the resource. Therefore, within the inner loop you _can_ check both this ingredient input _and_ what craft it's being checked for. The problem isn't that your suggestion is technically impossible, the problem is that it introduces too many new things to keep track of and every one of those is a new place for the script to fail. The goal is _simplicity_ - do as much as you can with as little as possible, because otherwise the chances of total failure rise rapidly, and if any per-tick operations are too complex the game _will_ die because you're trying to do too much at once - but your suggestion introduces _major_ complexity. The more complex it is, the more problems it'll have.

I definitely agree this is a big issue, but I haven't found a good way to _solve_ it yet...

**Comment by Quizer85** [comment]
> You'd loop _every_ resource in the game, by doing that. This loop only examines the inputs for the current recipe, your implementation would have to look at every resource in the game, for every crafting recipe in the game.

Uh, no, that's not really what I was describing. My bit of changed pseudocode is still in the inner loop, after the relevant ingredients have been already been fetched. Sorry for failing to format that in a way that would have made that clear.

Also, what I mean by `INGREDIENT matters / doesn't matter` is only whether that ingredient imposes constraints on being allowed to craft the currently examined recipe, due to user settings. Ingredients that are not part of the recipe don't even enter into it.
.

Anyway, perhaps this should be lowered in priority. I think I'd rather have tick-based throttling on every recipe rather than percentage thresholds on every non-storage-limited resource, if I had to choose. That would still allow implementing autocrafting of non-storage-limited resources in a useful way.

Also, I'd probably not try to add all possible recipes at once, but maybe start with the basic ones and work my way up to the more advanced recipes gradually. Let's say we go "autocrafting scaffolds would be useful since trading with spiders consumes them, so this would let us replace those losses" and implement that. Then we go "Okay, now that we have scaffold autocrafting, we can..." and work our way forward like that.

**Comment by VariableVixen** [comment]
Ah! Okay, I read `does not matter` as `isn't used here`, so I thought that was a replacement for the _full_ loop.

Okay, so tick throttling is more useful, you think? I'll see about setting that up then, we'll see how it goes. That one will be... _relatively_ easy to do, and won't conflict with this particular method if I implement both - tick checking will be in the outer recipe loop and skip the whole craft if they fail, ingredient thresholds go in the inner loop.

**Comment by Quizer85** [comment]
Yeah, I do think tick throttling will be very useful. It will grant great control over how fast certain resources are being consumed, which is going to be useful especially for alloy crafting, fur product crafting and any purely non-storage-limited recipes that are probably going to be added at some point:


- Alloy crafting: Alloy is the most lopsided of the currently supported recipes to a ridiculous degree. Titanium income in the late game is effectively arbitrarily high due to trade, but even now at the end of my third run I've only just barely reached the point where I can craft 1 batch of alloy per tick without rapidly chewing through my steel stockpile. Alloy autocrafting is going to be necessary to support eludium autocrafting in the space age, so it'll be good to have a level of control over the crafting speed.

- Fur product crafting: the currently available automation is good at using the input resources where possible, but enabling it usually results in having no stockpile of at least one part of the crafting chain. Some more granularity is going to be quite helpful here as well.
(This will also let us dispense with the weird and clunky way the fur product line is currently being automated and bring it in line with the rest of the automated recipes, though it can keep having its own section in the UI.)

- Non-storage-limited crafting: With the ability to freely select the crafting interval, there does not even strictly need to be a way to look at input resource levels to have useful automation. The user can just throttle the crafting enough to let both input and output resource grow.


All in all, this will add a ton of flexibility to the current level of automation and pave the way for future improvements.

**Comment by VariableVixen** [comment]
> Yeah, I do think tick throttling will be very useful.

I've managed to implement that, but the individualised thresholds are gonna be a longer time coming, I'm afraid.
